cpp/src/arrow/compute/function.cc

```diff
@@ -167,6 +180,109 @@ const Kernel* DispatchExactImpl(const Function* func,
   return nullptr;
 }
 
+struct FunctionExecutorImpl : public FunctionExecutor {
+  FunctionExecutorImpl(std::vector<TypeHolder> in_types, const Kernel* kernel,
+                       std::unique_ptr<detail::KernelExecutor> executor,
+                       const Function& func)
+      : in_types(std::move(in_types)),
+        kernel(kernel),
+        kernel_ctx(),
+        executor(std::move(executor)),
+        func(func),
+        state(),
+        options(NULLPTR) {}
+  virtual ~FunctionExecutorImpl() {}
+
+  Status KernelInit(const FunctionOptions* options) {
+    if (!kernel_ctx) {
+      return Status::Invalid("function executor with null kernel context");
+    }
+    if (options == nullptr) {
+      RETURN_NOT_OK(CheckOptions(func, options));
+      options = func.default_options();
+    }
```

Comment:
Could we fold this into 
```suggestion
    RETURN_NOT_OK(CheckOptions(func, &options));
```